pkg/controllers/sync/controller.go

```diff
@@ -1186,3 +1155,140 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
+
+		// get the corresponding federated object
+		fedObjName := naming.GenerateFederatedObjectName(clusterObj.GetName(), ftc.Name)
+		fedObj, err := s.fedAccessor.FederatedResource(common.QualifiedName{
+			Namespace: clusterObj.GetNamespace(),
+			Name:      fedObjName,
+		})
+		if err != nil {
+			keyedLogger.Error(err, "Failed to get corresponding federated object")
+			deleteFailedObjs = append(deleteFailedObjs, namespacedName)
+			continue
+		}
+		if fedObj == nil {
+			err := fmt.Errorf("the federated object %s is not found", fedObj.FederatedName())
+			keyedLogger.Error(err, "Failed to get corresponding federated object")
+			deleteFailedObjs = append(deleteFailedObjs, namespacedName)
+			continue
+		}
+
+		// execute the cascading-delete operation
+		dispatcher := dispatch.NewUnmanagedDispatcher(s.getClusterClient, fedObj.TargetGVR(), fedObj.TargetName(), s.metrics)
+		s.removeFromCluster(ctx, dispatcher, clusterName, fedObj, clusterObj, true)
+
+		dispatchOk, timeoutErr := dispatcher.Wait()
+		if timeoutErr != nil {
+			keyedLogger.Error(timeoutErr, "Cascading-delete cluster object timeout")
+		}
+		if !dispatchOk {
+			keyedLogger.Error(nil, "Failed to cascading-delete target object")
+			deleteFailedObjs = append(deleteFailedObjs, namespacedName)
+		}
+	}
+
+	// if the update operation fails, do not continue the delete operation,
+	// but return directly to avoid accidentally deleting any orphaned cluster objects.
+	if len(deleteFailedObjs) > 0 {
+		syncMap.Store(targetGVK, fmt.Sprintf("failed to update those cluster objects for %s: %v", targetGVK, deleteFailedObjs))
+	}
+}
+
+func (s *SyncController) listManagedClusterObjects(
+	ctx context.Context,
+	ftc *fedcorev1a1.FederatedTypeConfig,
+	clusterName string,
+) ([]*unstructured.Unstructured, error) {
+	clusterObjs := []*unstructured.Unstructured{}
+
+	resourceLister, hasSynced, exists := s.fedInformerManager.GetResourceLister(ftc.GetSourceTypeGVK(), clusterName)
+	if !exists {
+		return nil, fmt.Errorf("lister of cluster %s does not exist", clusterName)
+	}
+
+	// If cluster cache is synced, we check the store.
+	// Otherwise, we will have to issue a list request.
+	if hasSynced() {
+		objects, err := resourceLister.List(labels.Everything())
+		if err != nil {
+			return nil, fmt.Errorf("failed to list cluster objects in %s: %w", clusterName, err)
+		}
+
+		clusterObjs = make([]*unstructured.Unstructured, len(objects))
+		for i := range objects {
+			clusterObjs[i] = objects[i].(*unstructured.Unstructured)
+		}
+	} else {
+		client, err := s.getClusterClient(clusterName)
+		if err != nil {
+			return nil, err
+		}
+
+		objects, err := client.Resource(ftc.GetSourceTypeGVR()).Namespace(corev1.NamespaceAll).List(ctx,
+			metav1.ListOptions{
+				LabelSelector: labels.SelectorFromSet(labels.Set{
+					managedlabel.ManagedByKubeAdmiralLabelKey: managedlabel.ManagedByKubeAdmiralLabelValue,
+				}).String(),
```

Comment:
Should we list with ResourceVersion: "0"?

Reply:
I think we should. The code has been updated.